src/librustc_trans/back/write.rs

```diff
@@ -557,9 +561,15 @@ unsafe fn optimize_and_codegen(cgcx: &CodegenContext,
 
         if config.emit_obj {
             let path = output_names.with_extension(&format!("{}.o", name_extra));
-            with_codegen(tm, llmod, config.no_builtins, |cpm| {
-                write_output_file(cgcx.handler, tm, cpm, llmod, &path, llvm::ObjectFileType);
-            });
+            if !cgcx.is_like_pnacl {
+                with_codegen(tm, llmod, config.no_builtins, |cpm| {
```

Comment:
But there doesn't exist any PNaCl LLVM target machine in (`pnacl-`)LLVM's target registry (ie using the PNaCl LLVM fork doesn't help in any way). So there's no LLVM target to tell LLVM how to write an object file (of any format).

In an ideal world, I'd completely throw out the PNaCl IR simplification passes and restart afresh, and I'd do it with an actual, real LLVM target machine (since LLVM's ISD form and PNaCl's IR are actually pretty similar). Sadly, while LLVM's SelectionDAG does most (if not all) the heavy lifting w.r.t. legalizing IR to a specific target, the rest of the toolchain would still need modifications to know how to handle the new PNaCl object format (ie how to read/link), which I would have to create specifically for the target machine (another possible, but still non-trivial thing to implement). The required toolchain modifications are the reason JS also uses bitcode as its object format, despite having an official target machine in LLVM. Plus the translation from the object format back into LLVM IR for codegenning to the real native target.

While I grant that using bitcode as the object format is annoying at times, it is a lot easier because it integrates the preexisting LLVM gold plugin based linker.

I actually agree with you, but the impl cost of what you propose is pretty high vs what `emcc` and `pnacl-clang` currently do (ie `-emit-llvm` (both are just scripts wrapping the real `clang` underneath; no such luck for Rust)). Plus, the other toolchains would require breaking changes to conform (ie this logic makes Rust conform with the other toolchains).